lib/sda/node.go

```diff
@@ -0,0 +1,270 @@
+	instance ProtocolInstance
+	// aggregate messages in order to dispatch them at once in the protocol
+	// instance
+	msgQueue map[uuid.UUID][]*SDAData
+	// Holds flags that influence the behaviour of the node
+	flags uint32
+}
+
+const (
+	NCAggregateMessages = 1 << iota
+	NCTimeout           = 1 << iota
+)
+
+// MsgHandler is called upon reception of a certain message-type
+type MsgHandler func([]*interface{})
+
+// NewNode creates a new node
+func NewNode(o *Overlay, tok *Token) (*Node, error) {
+	n := &Node{overlay: o,
+		token:    tok,
+		channels: make(map[uuid.UUID]interface{}),
+		handlers: make(map[uuid.UUID]MsgHandler),
+		msgQueue: make(map[uuid.UUID][]*SDAData),
+		flags:    NCAggregateMessages,
+	}
+	return n, n.protocolInstantiate()
+}
+
+// TreeNode gets the treeNode of this node
+func (n *Node) TreeNode() *TreeNode {
+	tn, err := n.overlay.TreeNodeFromToken(n.token)
+	if err != nil {
+		dbg.Error("TreeNodeFromToken not found by token", err)
+		return nil
+	}
+	return tn
+}
+
+// Entity returns our entity
+func (n *Node) Entity() *network.Entity {
+	return n.TreeNode().Entity
+}
+
+// Parent returns the parent-TreeNode of ourselves
+func (n *Node) Parent() *TreeNode {
+	return n.TreeNode().Parent
+}
+
+// Children returns the children of ourselves
+func (n *Node) Children() []*TreeNode {
+	return n.TreeNode().Children
+}
+
+// Root returns the root-node of that tree
+func (n *Node) Root() *TreeNode {
+	return n.Tree().Root
+}
+
+// IsRoot returns whether whether we are at the top of the tree
+func (n *Node) IsRoot() bool {
+	return n.TreeNode().Parent == nil
+}
+
+// IsLeaf returns whether whether we are at the bottom of the tree
+func (n *Node) IsLeaf() bool {
+	return len(n.TreeNode().Children) == 0
+}
+
+// SendTo sends to a given node
+func (n *Node) SendTo(to *TreeNode, msg interface{}) error {
+	if to == nil {
+		return errors.New("Sent to a nil TreeNode")
+	}
+	return n.overlay.SendToTreeNode(n.token, to, msg)
+}
+
+// Tree returns the tree of that node
+func (n *Node) Tree() *Tree {
+	return n.overlay.TreeFromToken(n.token)
+}
+
+// EntityList returns the entity-list
+func (n *Node) EntityList() *EntityList {
+	return n.Tree().EntityList
+}
+
+// RegisterChannel takes a channel with a struct that contains two
+// elements: a TreeNode and a message.
+func (n *Node) RegisterChannel(c interface{}) error {
+	cr := reflect.TypeOf(c)
+	// Check we have the correct channel-type
+	if cr.Kind() != reflect.Chan {
+		return errors.New("Input is not channel")
+	}
+	if cr.Elem().Kind() != reflect.Struct {
+		return errors.New("Input is not channel of structure")
+	}
+	if cr.Elem().NumField() != 2 {
+		return errors.New("Input is not channel of structure with 2 elements")
+	}
+	dbg.Lvl3(cr.Elem().Field(0).Type)
+	if cr.Elem().Field(0).Type != reflect.TypeOf(TreeNode{}) {
+		return errors.New("Input-channel doesn't have TreeNode as element")
+	}
+	typ := network.RegisterMessageUUID(network.RTypeToUUID(cr.Elem().Field(1).Type),
```

Comment:
Do we need to register to the network ? That way, in host.go, the message will be decoded ?
But are we not already registering our type somewhere else ? Every message that applications uses are already supposed to be registered, no ?  I think you only want to get the type out of the message here and rely on network.RegisterMessageUUID which should not be used for that.  Or did I missed something ?